stdlib/mexpr/ast.mc

```diff
@@ -74,142 +85,187 @@ lang FunAst = VarAst + AppAst
+  | TmConst t -> TmConst t
 
   sem sfold_Expr_Expr (f : a -> b -> a) (acc : a) =
-  | TmType t -> f acc t.inexpr
+  | TmConst t -> acc
 end
 
-
-lang RecLetsAst = VarAst
+-- TmSeq --
+lang SeqAst
   syn Expr =
-  | TmRecLets {bindings : [{ident : Name,
-                            ty    : Type,
-                            body  : Expr}],
-               inexpr   : Expr,
-               ty       : Type}
+  | TmSeq {tms : [Expr],
+           ty: Type,
+           info: Info}
+
+  sem info =
+  | TmSeq r -> r.info
 
   sem ty =
-  | TmRecLets t -> t.ty
+  | TmSeq t -> t.ty
 
   sem withType (ty : Type) =
-  | TmRecLets t -> TmRecLets {t with ty = ty}
+  | TmSeq t -> TmSeq {t with ty = ty}
 
   sem smap_Expr_Expr (f : Expr -> a) =
-  | TmRecLets t ->
-    TmRecLets {{t with bindings = map (lam b. {b with body = f b.body})
-                                      t.bindings}
-                  with inexpr = f t.inexpr}
+  | TmSeq t -> TmSeq {t with tms = map f t.tms}
 
   sem sfold_Expr_Expr (f : a -> b -> a) (acc : a) =
-  | TmRecLets t -> f (foldl f acc (map (lam b. b.body) t.bindings)) t.inexpr
+  | TmSeq t -> foldl f acc t.tms
 end
 
-lang ConstAst
-  syn Const =
 
+-- TmRecord and TmRecorsUpdate -- 
+lang RecordAst
+  syn Expr =
+  | TmRecord {bindings : AssocMap String Expr,
+              ty : Type, 
+              info : Info}
+  | TmRecordUpdate {rec : Expr,
+                    key : String,
+                    value : Expr,
+                    ty : Type,
+                    info : Info}
+
+  sem info =
+  | TmRecord r -> r.info
+  | TmRecordUpdate r -> r.info
+
+  sem ty =
+  | TmRecord t -> t.ty
+  | TmRecordUpdate t -> t.ty
+
+  sem withType (ty : Type) =
+  | TmRecord t -> TmRecord {t with ty = ty}
+  | TmRecordUpdate t -> TmRecordUpdate {t with ty = ty}
+
+  sem smap_Expr_Expr (f : Expr -> a) =
+  | TmRecord t -> TmRecord {t with bindings = assocMap {eq=eqString} f t.bindings}
+  | TmRecordUpdate t -> TmRecordUpdate {{t with rec = f t.rec}
+                                           with value = f t.value}
+
+  sem sfold_Expr_Expr (f : a -> b -> a) (acc : a) =
+  | TmRecord t -> assocFold {eq=eqString} (lam acc. lam _k. lam v. f acc v) acc t.bindings
+  | TmRecordUpdate t -> f (f acc t.rec) t.value
+end
+
+-- TmType -- 
+lang TypeAst
   syn Expr =
-  | TmConst {val : Const, ty: Type, fi: Info}
+  | TmType {ident : Name,
+            ty : Type,
+            inexpr : Expr,
+            info : Info}
 
   sem info =
-  | TmConst r -> r.fi
+  | TmType r -> r.info
 
   sem ty =
-  | TmConst t -> t.ty
+  | TmType t -> t.ty
 
   sem withType (ty : Type) =
-  | TmConst t -> TmConst {t with ty = ty}
+  | TmType t -> TmType {t with ty = ty}
 
   sem smap_Expr_Expr (f : Expr -> a) =
-  | TmConst t -> TmConst t
+  | TmType t -> TmType {t with inexpr = f t.inexpr}
 
   sem sfold_Expr_Expr (f : a -> b -> a) (acc : a) =
-  | TmConst t -> acc
+  | TmType t -> f acc t.inexpr
 end
 
+-- TmCondef and TmConApp --
```

Comment:
Inconsistent capitalization on `TmConDef` in the comment.
```suggestion
-- TmConDef and TmConApp --
```

Reply:
Fixed!

Reply:
I changed the name to TmConDef and TmConApp in boot, so that it is consistent now.